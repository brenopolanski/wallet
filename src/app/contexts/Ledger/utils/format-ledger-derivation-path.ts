import { LedgerDerivationScheme } from "app/contexts/Ledger/contracts";

export const formatLedgerDerivationPath = (scheme: LedgerDerivationScheme) =>
	`m/${scheme.purpose || 44}'/${scheme.coinType}'/${scheme.account || 0}'/${scheme.change || 0}/${
		scheme.address || 0
	}`;
