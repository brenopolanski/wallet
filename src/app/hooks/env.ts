import { Networks } from "@payvo/sdk";
import { sortBy } from "@payvo/sdk-helpers";
import { Contracts } from "@payvo/sdk-profiles";
import { useEnvironmentContext } from "app/contexts/Environment";
import { useMemo } from "react";
import { useHistory, useParams } from "react-router-dom";

export const useActiveProfile = (): Contracts.IProfile => {
	const history = useHistory();

	const context = useEnvironmentContext();
	const { profileId } = useParams<{ profileId: string }>();

	return useMemo(() => {
		if (!profileId) {
			throw new Error(
				`Parameter [profileId] must be available on the route where [useActiveProfile] is called. Current route is [${history.location.pathname}].`,
			);
		}

		return context.env.profiles().findById(profileId);
	}, [context.env, history.location.pathname, profileId]);
};

export const useActiveWallet = (): Contracts.IReadWriteWallet => {
	const profile = useActiveProfile();
	const { walletId } = useParams<{ walletId: string }>();

	return useMemo(() => profile.wallets().findById(walletId), [profile, walletId]);
};

export const useNetworks = (profile: Contracts.IProfile) =>
	useMemo<Networks.Network[]>(() => {
		const results = profile
			.wallets()
			.values()
			.reduce<Record<string, Networks.Network>>(
				(accumulator, wallet) => ({
					...accumulator,
					[wallet.networkId()]: wallet.network(),
				}),
				{},
			);

		return sortBy(Object.values(results), (network) => network.displayName());
	}, [profile]);

export const useActiveWalletWhenNeeded = (isRequired: boolean) => {
	const profile = useActiveProfile();
	const { walletId } = useParams<{ walletId: string }>();

	return useMemo(() => {
		try {
			return profile.wallets().findById(walletId);
		} catch (error) {
			if (isRequired) {
				throw error;
			}

			return undefined;
		}
	}, [isRequired, profile, walletId]);
};
