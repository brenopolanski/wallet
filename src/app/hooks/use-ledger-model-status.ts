import { WalletLedgerModel } from "@payvo/sdk-profiles/distribution/contracts";
import { useMemo } from "react";

export type LedgerModel = WalletLedgerModel.NanoS | WalletLedgerModel.NanoX;

interface ModelStatusProperties {
	connectedModel?: LedgerModel;
	supportedModels?: LedgerModel[];
}

export const useLedgerModelStatus = ({
	connectedModel,
	supportedModels = [WalletLedgerModel.NanoS, WalletLedgerModel.NanoX],
}: ModelStatusProperties) => {
	const isLedgerModelSupported = useMemo(() => {
		if (!connectedModel) {
			return false;
		}

		return supportedModels.includes(connectedModel);
	}, [connectedModel, supportedModels]);

	return { isLedgerModelSupported };
};
