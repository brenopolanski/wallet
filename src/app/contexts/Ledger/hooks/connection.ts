import LedgerTransportNodeHID from "@ledgerhq/hw-transport-node-hid-singleton";
import { Coins } from "@payvo/sdk";
import { Contracts } from "@payvo/sdk-profiles";
import { WalletData } from "@payvo/sdk-profiles/distribution/contracts";
import { useEnvironmentContext } from "app/contexts/Environment";
import { LedgerData, minVersionList } from "app/contexts/Ledger/contracts";
import { formatLedgerDerivationPath } from "app/contexts/Ledger/utils/format-ledger-derivation-path";
import { toasts } from "app/services";
import retry from "async-retry";
import { getDefaultAlias } from "domains/wallet/utils/get-default-alias";
import { useCallback, useEffect, useMemo, useReducer, useRef, useState } from "react";
import { useTranslation } from "react-i18next";
import semver from "semver";

import { connectionReducer, defaultConnectionState } from "./connection.state";

export const useLedgerConnection = (transport: typeof LedgerTransportNodeHID) => {
	const { t } = useTranslation();

	const { persist } = useEnvironmentContext();
	const [state, dispatch] = useReducer(connectionReducer, defaultConnectionState);
	const abortRetryReference = useRef<boolean>(false);

	const [deviceName, setDeviceName] = useState<string | undefined>();

	const { isBusy, isConnected, error } = state;

	useEffect(() => {
		if (deviceName) {
			if (isConnected) {
				toasts.success(t("COMMON.LEDGER_CONNECTED", { device: deviceName }));
			} else {
				toasts.warning(t("COMMON.LEDGER_DISCONNECTED", { device: deviceName }));
			}
		}
	}, [isConnected, t]); // eslint-disable-line react-hooks/exhaustive-deps

	const listenDevice = useCallback(
		() =>
			transport.listen({
				complete: () => void 0,

				error: (e) => dispatch({ message: e.message, type: "failed" }),
				// @ts-ignore
				next: ({ type, descriptor, deviceModel }) => {
					setDeviceName(deviceModel?.productName);

					if (type === "add") {
						dispatch({ id: deviceModel?.id || "nanoS", path: descriptor, type: "add" });
						return;
					}

					dispatch({ type: "remove" });
				},
			}),
		[transport],
	);

	const importLedgerWallets = useCallback(
		async (wallets: LedgerData[], coin: Coins.Coin, profile: Contracts.IProfile) => {
			for (const { address, path } of wallets) {
				const wallet = await profile.walletFactory().fromAddressWithDerivationPath({
					address,
					coin: coin.network().coin(),
					network: coin.network().id(),
					path,
				});

				profile.wallets().push(wallet);

				wallet.mutator().alias(
					getDefaultAlias({
						network: wallet.network(),
						profile,
					}),
				);

				wallet.data().set(WalletData.LedgerModel, state.device?.id);
			}
			await persist();
		},
		[persist, state],
	);

	const connect = useCallback(
		async (
			profile: Contracts.IProfile,
			coin: string,
			network: string,
			retryOptions: retry.Options = { factor: 1, randomize: false, retries: 50 },
		) => {
			dispatch({ type: "waiting" });
			abortRetryReference.current = false;

			const instance = profile.coins().set(coin, network);

			try {
				const slip44 = instance.config().get<number>("network.constants.slip44");

				const connectFunction: retry.RetryFunction<void> = async (bail, attempts) => {
					if (abortRetryReference.current && attempts > 1) {
						bail(new Error(t("WALLETS.MODAL_LEDGER_WALLET.GENERIC_CONNECTION_ERROR")));
					}

					await instance.__construct();
					await instance.ledger().connect(transport);

					if (minVersionList[coin]) {
						const currentVersion = await instance.ledger().getVersion();

						if (semver.lt(currentVersion, minVersionList[coin])) {
							bail(
								new Error(
									t("WALLETS.MODAL_LEDGER_WALLET.UPDATE_ERROR", { coin, version: currentVersion }),
								),
							);
						}
					}

					// Ensure that the app is accessible
					await instance.ledger().getPublicKey(formatLedgerDerivationPath({ coinType: slip44 }));
				};

				await retry(connectFunction, retryOptions);
				dispatch({ type: "connected" });
			} catch (connectError) {
				const errorMessage =
					connectError.statusText === "UNKNOWN_ERROR"
						? t("WALLETS.MODAL_LEDGER_WALLET.GENERIC_CONNECTION_ERROR")
						: connectError.message;

				try {
					await instance.ledger().disconnect();
				} catch {
					//
				}
				dispatch({ message: errorMessage, type: "failed" });
				throw connectError;
			}
		},
		[t, transport],
	);

	const disconnect = useCallback(async (coin: Coins.Coin) => {
		await coin.ledger().disconnect();
		dispatch({ type: "disconnected" });
	}, []);

	const setBusy = useCallback(() => dispatch({ type: "busy" }), []);
	const setIdle = useCallback(() => dispatch({ type: "connected" }), []);

	const abortConnectionRetry = useCallback(() => (abortRetryReference.current = true), []);
	const isAwaitingConnection = useMemo(() => state.isWaiting && !state.isConnected, [state]);
	const isAwaitingDeviceConfirmation = useMemo(() => state.isWaiting && state.isConnected, [state]);
	const hasDeviceAvailable = useMemo(() => !!state.device, [state]);

	useEffect(() => {
		const subscription = listenDevice();
		return () => {
			subscription?.unsubscribe();
		};
	}, [listenDevice]);

	return {
		abortConnectionRetry,
		connect,
		disconnect,
		dispatch,
		error,
		hasDeviceAvailable,
		importLedgerWallets,
		isAwaitingConnection,
		isAwaitingDeviceConfirmation,
		isBusy,
		isConnected,
		ledgerDevice: state.device,
		setBusy,
		setIdle,
		transport,
	};
};
