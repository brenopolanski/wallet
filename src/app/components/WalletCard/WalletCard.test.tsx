import { Contracts } from "@payvo/profiles";
import userEvent from "@testing-library/user-event";
import * as useRandomNumberHook from "app/hooks/use-random-number";
import { createMemoryHistory } from "history";
import React from "react";
import { Route } from "react-router-dom";
import { env, getDefaultProfileId, render, screen } from "utils/testing-library";

import { WalletCard } from "./WalletCard";

const dashboardURL = `/profiles/${getDefaultProfileId()}/dashboard`;
const history = createMemoryHistory();

let profile: Contracts.IProfile;
let wallet: Contracts.IReadWriteWallet;

describe("Wallet Card", () => {
	beforeAll(() => {
		jest.spyOn(useRandomNumberHook, "useRandomNumber").mockImplementation(() => 1);
	});

	beforeEach(async () => {
		profile = env.profiles().findById(getDefaultProfileId());
		wallet = profile.wallets().findById("ac38fe6d-4b67-4ef1-85be-17c5f6841129");
		wallet.data().set(Contracts.WalletFlag.Starred, true);
		wallet.data().set(Contracts.WalletData.DerivationPath, "0");

		await wallet.synchroniser().identity();
		jest.spyOn(wallet, "isMultiSignature").mockReturnValue(true);

		history.push(dashboardURL);
	});

	afterAll(() => {
		useRandomNumberHook.useRandomNumber.mockRestore();
	});

	it("should render", () => {
		const { container } = render(
			<Route path="/profiles/:profileId/dashboard">
				<WalletCard wallet={wallet} />
			</Route>,
			{
				history,
				routes: [dashboardURL],
			},
		);

		expect(container).toMatchSnapshot();
	});

	it("should render loading state", () => {
		const { container } = render(
			<Route path="/profiles/:profileId/dashboard">
				<WalletCard isLoading={true} wallet={wallet} />
			</Route>,
			{
				history,
				routes: [dashboardURL],
			},
		);

		expect(screen.getByTestId("WalletCard__skeleton")).toBeInTheDocument();
		expect(container).toMatchSnapshot();
	});

	it("should render blank", () => {
		const { container } = render(
			<Route path="/profiles/:profileId/dashboard">
				<WalletCard />
			</Route>,
			{
				history,
				routes: [dashboardURL],
			},
		);

		expect(container).toMatchSnapshot();
	});

	it("should render blank with starred display type", () => {
		const { container } = render(
			<Route path="/profiles/:profileId/dashboard">
				<WalletCard displayType="starred" />
			</Route>,
			{
				history,
				routes: [dashboardURL],
			},
		);

		expect(container).toHaveTextContent("star-filled.svg");
		expect(container).toMatchSnapshot();
	});

	it("should render blank with ledger display type", () => {
		const { container } = render(
			<Route path="/profiles/:profileId/dashboard">
				<WalletCard displayType="ledger" />
			</Route>,
			{
				history,
				routes: [dashboardURL],
			},
		);

		expect(container).toHaveTextContent("ledger.svg");
		expect(container).toMatchSnapshot();
	});

	it("should render with wallet data", () => {
		wallet.settings().set(Contracts.WalletSetting.Alias, "My wallet");

		const { container } = render(
			<Route path="/profiles/:profileId/dashboard">
				<WalletCard wallet={wallet} />,
			</Route>,
			{
				history,
				routes: [dashboardURL],
			},
		);

		expect(container).toMatchSnapshot();
	});

	it("should render with wallet data and optional icon", () => {
		wallet.settings().set(Contracts.WalletSetting.Alias, "My wallet");

		const { container } = render(
			<Route path="/profiles/:profileId/dashboard">
				<WalletCard wallet={wallet} />
			</Route>,
			{
				history,
				routes: [dashboardURL],
			},
		);

		expect(container).toMatchSnapshot();
	});

	it("should click a wallet and redirect to it when fully restored", () => {
		render(
			<Route path="/profiles/:profileId/dashboard">
				<WalletCard wallet={wallet} />
			</Route>,
			{
				history,
				routes: [dashboardURL],
			},
		);

		expect(history.location.pathname).toBe(`/profiles/${profile.id()}/dashboard`);

		userEvent.click(screen.getByText("D8rr7B1d6TL6pf14LgMz4sKp1VBMs6YUYD"));

		expect(history.location.pathname).toBe(`/profiles/${profile.id()}/wallets/${wallet.id()}`);
	});

	it("should not redirect to wallet when not fully restored", () => {
		const hasBeenFullyRestored = jest.spyOn(wallet, "hasBeenFullyRestored").mockReturnValue(false);

		render(
			<Route path="/profiles/:profileId/dashboard">
				<WalletCard wallet={wallet} />
			</Route>,
			{
				history,
				routes: [dashboardURL],
			},
		);

		expect(history.location.pathname).toBe(`/profiles/${profile.id()}/dashboard`);

		userEvent.click(screen.getByText(wallet.alias()!));

		expect(history.location.pathname).not.toBe(`/profiles/${profile.id()}/wallets/${wallet.id()}`);

		hasBeenFullyRestored.mockRestore();
	});

	it("should execute onWalletAction callback", () => {
		const onWalletAction = jest.fn();
		const actions = [{ label: "Option", value: "option" }];

		render(
			<Route path="/profiles/:profileId/dashboard">
				<WalletCard wallet={wallet} actions={actions} onWalletAction={onWalletAction} />
			</Route>,
			{
				history,
				routes: [dashboardURL],
			},
		);

		expect(history.location.pathname).toBe(`/profiles/${profile.id()}/dashboard`);

		userEvent.click(screen.getByTestId("dropdown__toggle"));

		expect(screen.getByTestId("dropdown__content")).toBeInTheDocument();
		expect(screen.getByTestId("dropdown__option--0")).toBeInTheDocument();

		userEvent.click(screen.getByTestId("dropdown__option--0"));

		expect(onWalletAction).toHaveBeenCalledWith(actions[0].value, wallet);
	});
});
