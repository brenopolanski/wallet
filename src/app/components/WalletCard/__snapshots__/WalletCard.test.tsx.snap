// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Wallet Card should render 1`] = `
<div>
  <div
    class="w-64 inline-block"
    data-testid="WalletCard__D8rr7B1d6TL6pf14LgMz4sKp1VBMs6YUYD"
  >
    <div
      class="h-48"
    >
      <button
        class="Cardstyles__CardButton-sc-ajtauu-0 dZiOWL"
        data-testid="Card"
        type="button"
      >
        <div
          class="flex relative flex-col justify-between p-5 h-full"
        >
          <div
            class="flex items-center space-x-4"
          >
            <div
              class="-space-x-1 whitespace-nowrap"
            >
              <div
                aria-label="ARK Devnet"
                class="Circle__CircleWrapper-sc-d5u41q-0 egSsGQ text-theme-secondary-700 border-theme-secondary-300 dark:border-theme-secondary-700 ring-theme-background"
                data-testid="NetworkIcon-ARK-ark.devnet"
              >
                <div
                  class="sc-bdvvtL qGXhc"
                  data-testid="NetworkIcon__icon"
                  height="20"
                  width="20"
                >
                  <svg>
                    ark.svg
                  </svg>
                </div>
              </div>
              <div
                class="Avatar__AvatarWrapper-sc-18vw8a4-0 ewMeZp flex-shrink-0"
                data-testid="Avatar"
              >
                <div
                  class="w-full h-full inline-flex items-center justify-center overflow-hidden align-middle rounded-full"
                >
                  <img
                    alt="D8rr7B1d6TL6pf14LgMz4sKp1VBMs6YUYD"
                    src="data:image/svg+xml;utf8,<svg version=\\"1.1\\" xmlns=\\"http://www.w3.org/2000/svg\\" class=\\"picasso\\" width=\\"100\\" height=\\"100\\" viewBox=\\"0 0 100 100\\"><style>.picasso circle{mix-blend-mode:soft-light;}</style><rect fill=\\"rgb(0, 150, 136)\\" width=\\"100\\" height=\\"100\\"/><circle r=\\"50\\" cx=\\"80\\" cy=\\"50\\" fill=\\"rgb(76, 175, 80)\\"/><circle r=\\"35\\" cx=\\"100\\" cy=\\"40\\" fill=\\"rgb(156, 39, 176)\\"/><circle r=\\"40\\" cx=\\"90\\" cy=\\"70\\" fill=\\"rgb(3, 169, 244)\\"/></svg>"
                    title="D8rr7B1d6TL6pf14LgMz4sKp1VBMs6YUYD"
                  />
                </div>
              </div>
            </div>
            <span
              class="font-semibold truncate text-theme-secondary-text"
            >
              ARK Wallet 1
            </span>
          </div>
          <span
            class="whitespace-nowrap mt-auto text-lg font-bold text-theme-text"
            data-testid="Amount"
          >
            33.75089801 DARK
          </span>
          <span
            class="inline-flex overflow-hidden mt-1 text-xs font-semibold whitespace-nowrap text-theme-secondary-text no-ligatures"
          >
            D8rr7B1d6TL6pf14LgMz4sKp1VBMs6YUYD
          </span>
        </div>
        <div
          class="flex absolute -top-1 -right-1 items-center m-4 space-x-1"
        >
          <div
            class="inline-block p-1 text-theme-secondary-700"
            data-testid="WalletIcon__Ledger"
          >
            <div
              class="sc-bdvvtL qGXhc"
              height="20"
              width="20"
            >
              <svg>
                ledger.svg
              </svg>
            </div>
          </div>
          <div
            class="inline-block p-1 text-theme-warning-400"
            data-testid="WalletIcon__Starred"
          >
            <div
              class="sc-bdvvtL qGXhc"
              height="20"
              width="20"
            >
              <svg>
                star-filled.svg
              </svg>
            </div>
          </div>
          <div
            class="inline-block p-1 text-theme-secondary-700"
            data-testid="WalletIcon__Multisignature"
          >
            <div
              class="sc-bdvvtL qGXhc"
              height="20"
              width="20"
            >
              <svg>
                multisignature.svg
              </svg>
            </div>
          </div>
          <div
            class="inline-block p-1 text-theme-secondary-700"
            data-testid="WalletIcon__TestNetwork"
          >
            <div
              class="sc-bdvvtL qGXhc"
              height="20"
              width="20"
            >
              <svg>
                code.svg
              </svg>
            </div>
          </div>
        </div>
      </button>
    </div>
  </div>
</div>
`;

exports[`Wallet Card should render blank 1`] = `
<div>
  <div
    class="w-64 inline-block"
    data-testid="WalletCard__blank"
  >
    <div
      class="h-48"
    >
      <button
        class="Cardstyles__CardButton-sc-ajtauu-0 gawQiQ"
        data-testid="Card"
        tabindex="-1"
        type="button"
      >
        <div
          class="flex flex-col justify-between p-5 h-full"
        >
          <div
            class="flex -space-x-1"
          >
            <div
              class="Circle__CircleWrapper-sc-d5u41q-0 egSsGQ bg-theme-background border-theme-primary-100 dark:border-theme-secondary-800 ring-theme-background"
            />
            <div
              class="Circle__CircleWrapper-sc-d5u41q-0 egSsGQ bg-theme-background border-theme-primary-100 dark:border-theme-secondary-800 ring-theme-background"
            />
          </div>
          <div
            class="mt-auto text-lg font-bold text-theme-primary-100 dark:text-theme-secondary-800"
          >
            Balance
          </div>
          <span
            class="mt-1 text-xs font-semibold truncate text-theme-primary-100 dark:text-theme-secondary-800"
          >
            Address
          </span>
        </div>
        <div
          class="flex absolute -top-1 -right-1 items-center m-4 space-x-1"
        />
      </button>
    </div>
  </div>
</div>
`;

exports[`Wallet Card should render blank with ledger display type 1`] = `
<div>
  <div
    class="w-64 inline-block"
    data-testid="WalletCard__blank"
  >
    <div
      class="h-48"
    >
      <button
        class="Cardstyles__CardButton-sc-ajtauu-0 gawQiQ"
        data-testid="Card"
        tabindex="-1"
        type="button"
      >
        <div
          class="flex flex-col justify-between p-5 h-full"
        >
          <div
            class="flex -space-x-1"
          >
            <div
              class="Circle__CircleWrapper-sc-d5u41q-0 egSsGQ bg-theme-background border-theme-primary-100 dark:border-theme-secondary-800 ring-theme-background"
            />
            <div
              class="Circle__CircleWrapper-sc-d5u41q-0 egSsGQ bg-theme-background border-theme-primary-100 dark:border-theme-secondary-800 ring-theme-background"
            />
          </div>
          <div
            class="mt-auto text-lg font-bold text-theme-primary-100 dark:text-theme-secondary-800"
          >
            Balance
          </div>
          <span
            class="mt-1 text-xs font-semibold truncate text-theme-primary-100 dark:text-theme-secondary-800"
          >
            Address
          </span>
        </div>
        <div
          class="flex absolute -top-1 -right-1 items-center m-4 space-x-1"
        >
          <div
            class="sc-bdvvtL fCrhkg text-theme-primary-100 dark:text-theme-secondary-800"
            height="16"
            width="16"
          >
            <svg>
              ledger.svg
            </svg>
          </div>
        </div>
      </button>
    </div>
  </div>
</div>
`;

exports[`Wallet Card should render blank with starred display type 1`] = `
<div>
  <div
    class="w-64 inline-block"
    data-testid="WalletCard__blank"
  >
    <div
      class="h-48"
    >
      <button
        class="Cardstyles__CardButton-sc-ajtauu-0 gawQiQ"
        data-testid="Card"
        tabindex="-1"
        type="button"
      >
        <div
          class="flex flex-col justify-between p-5 h-full"
        >
          <div
            class="flex -space-x-1"
          >
            <div
              class="Circle__CircleWrapper-sc-d5u41q-0 egSsGQ bg-theme-background border-theme-primary-100 dark:border-theme-secondary-800 ring-theme-background"
            />
            <div
              class="Circle__CircleWrapper-sc-d5u41q-0 egSsGQ bg-theme-background border-theme-primary-100 dark:border-theme-secondary-800 ring-theme-background"
            />
          </div>
          <div
            class="mt-auto text-lg font-bold text-theme-primary-100 dark:text-theme-secondary-800"
          >
            Balance
          </div>
          <span
            class="mt-1 text-xs font-semibold truncate text-theme-primary-100 dark:text-theme-secondary-800"
          >
            Address
          </span>
        </div>
        <div
          class="flex absolute -top-1 -right-1 items-center m-4 space-x-1"
        >
          <div
            class="sc-bdvvtL fCrhkg text-theme-primary-100 dark:text-theme-secondary-800"
            height="16"
            width="16"
          >
            <svg>
              star-filled.svg
            </svg>
          </div>
        </div>
      </button>
    </div>
  </div>
</div>
`;

exports[`Wallet Card should render loading state 1`] = `
<div>
  <div
    class="bg-theme-background"
  >
    <button
      class="Cardstyles__CardButton-sc-ajtauu-0 gawQiQ"
      data-testid="Card"
      tabindex="-1"
      type="button"
    >
      <div
        class="flex relative flex-col justify-between p-5 h-full"
        data-testid="WalletCard__skeleton"
      >
        <div
          class="flex items-center"
        >
          <div
            class="flex items-center mr-4 -space-x-1"
          >
            <div
              class="Circle__CircleWrapper-sc-d5u41q-0 egSsGQ border-transparent ring-theme-background"
            >
              <span>
                <span
                  class="flex items-center leading-none"
                >
                  <span
                    class="react-loading-skeleton css-1eg0sfo-skeletonStyles-Skeleton"
                    style="width: 44px; height: 44px; border-radius: 50%;"
                  >
                    ‌
                  </span>
                  ‌
                </span>
              </span>
            </div>
            <div
              class="Circle__CircleWrapper-sc-d5u41q-0 egSsGQ border-transparent ring-theme-background"
            >
              <span>
                <span
                  class="flex items-center leading-none"
                >
                  <span
                    class="react-loading-skeleton css-1eg0sfo-skeletonStyles-Skeleton"
                    style="width: 44px; height: 44px; border-radius: 50%;"
                  >
                    ‌
                  </span>
                  ‌
                </span>
              </span>
            </div>
          </div>
          <span>
            <span
              class="flex items-center leading-none"
            >
              <span
                class="react-loading-skeleton css-1eg0sfo-skeletonStyles-Skeleton"
                style="width: 1px; height: 16px;"
              >
                ‌
              </span>
              ‌
            </span>
          </span>
        </div>
        <div
          class="flex items-center mt-4 space-x-1 h-7"
        >
          <span>
            <span
              class="flex items-center leading-none"
            >
              <span
                class="react-loading-skeleton css-1eg0sfo-skeletonStyles-Skeleton"
                style="width: 1px; height: 18px;"
              >
                ‌
              </span>
              ‌
            </span>
          </span>
          <span>
            <span
              class="flex items-center leading-none"
            >
              <span
                class="react-loading-skeleton css-1eg0sfo-skeletonStyles-Skeleton"
                style="width: 35px; height: 18px;"
              >
                ‌
              </span>
              ‌
            </span>
          </span>
        </div>
        <span
          class="flex items-center mt-1 h-4"
        >
          <span>
            <span
              class="flex items-center leading-none"
            >
              <span
                class="react-loading-skeleton css-1eg0sfo-skeletonStyles-Skeleton"
                style="width: 1px; height: 12px;"
              >
                ‌
              </span>
              ‌
            </span>
          </span>
        </span>
      </div>
      <div
        class="flex absolute -top-1 -right-1 items-center m-4 space-x-1"
      />
    </button>
  </div>
</div>
`;

exports[`Wallet Card should render with wallet data 1`] = `
<div>
  <div
    class="w-64 inline-block"
    data-testid="WalletCard__D8rr7B1d6TL6pf14LgMz4sKp1VBMs6YUYD"
  >
    <div
      class="h-48"
    >
      <button
        class="Cardstyles__CardButton-sc-ajtauu-0 dZiOWL"
        data-testid="Card"
        type="button"
      >
        <div
          class="flex relative flex-col justify-between p-5 h-full"
        >
          <div
            class="flex items-center space-x-4"
          >
            <div
              class="-space-x-1 whitespace-nowrap"
            >
              <div
                aria-label="ARK Devnet"
                class="Circle__CircleWrapper-sc-d5u41q-0 egSsGQ text-theme-secondary-700 border-theme-secondary-300 dark:border-theme-secondary-700 ring-theme-background"
                data-testid="NetworkIcon-ARK-ark.devnet"
              >
                <div
                  class="sc-bdvvtL qGXhc"
                  data-testid="NetworkIcon__icon"
                  height="20"
                  width="20"
                >
                  <svg>
                    ark.svg
                  </svg>
                </div>
              </div>
              <div
                class="Avatar__AvatarWrapper-sc-18vw8a4-0 ewMeZp flex-shrink-0"
                data-testid="Avatar"
              >
                <div
                  class="w-full h-full inline-flex items-center justify-center overflow-hidden align-middle rounded-full"
                >
                  <img
                    alt="D8rr7B1d6TL6pf14LgMz4sKp1VBMs6YUYD"
                    src="data:image/svg+xml;utf8,<svg version=\\"1.1\\" xmlns=\\"http://www.w3.org/2000/svg\\" class=\\"picasso\\" width=\\"100\\" height=\\"100\\" viewBox=\\"0 0 100 100\\"><style>.picasso circle{mix-blend-mode:soft-light;}</style><rect fill=\\"rgb(0, 150, 136)\\" width=\\"100\\" height=\\"100\\"/><circle r=\\"50\\" cx=\\"80\\" cy=\\"50\\" fill=\\"rgb(76, 175, 80)\\"/><circle r=\\"35\\" cx=\\"100\\" cy=\\"40\\" fill=\\"rgb(156, 39, 176)\\"/><circle r=\\"40\\" cx=\\"90\\" cy=\\"70\\" fill=\\"rgb(3, 169, 244)\\"/></svg>"
                    title="D8rr7B1d6TL6pf14LgMz4sKp1VBMs6YUYD"
                  />
                </div>
              </div>
            </div>
            <span
              class="font-semibold truncate text-theme-secondary-text"
            >
              My wallet
            </span>
          </div>
          <span
            class="whitespace-nowrap mt-auto text-lg font-bold text-theme-text"
            data-testid="Amount"
          >
            33.75089801 DARK
          </span>
          <span
            class="inline-flex overflow-hidden mt-1 text-xs font-semibold whitespace-nowrap text-theme-secondary-text no-ligatures"
          >
            D8rr7B1d6TL6pf14LgMz4sKp1VBMs6YUYD
          </span>
        </div>
        <div
          class="flex absolute -top-1 -right-1 items-center m-4 space-x-1"
        >
          <div
            class="inline-block p-1 text-theme-secondary-700"
            data-testid="WalletIcon__Ledger"
          >
            <div
              class="sc-bdvvtL qGXhc"
              height="20"
              width="20"
            >
              <svg>
                ledger.svg
              </svg>
            </div>
          </div>
          <div
            class="inline-block p-1 text-theme-warning-400"
            data-testid="WalletIcon__Starred"
          >
            <div
              class="sc-bdvvtL qGXhc"
              height="20"
              width="20"
            >
              <svg>
                star-filled.svg
              </svg>
            </div>
          </div>
          <div
            class="inline-block p-1 text-theme-secondary-700"
            data-testid="WalletIcon__Multisignature"
          >
            <div
              class="sc-bdvvtL qGXhc"
              height="20"
              width="20"
            >
              <svg>
                multisignature.svg
              </svg>
            </div>
          </div>
          <div
            class="inline-block p-1 text-theme-secondary-700"
            data-testid="WalletIcon__TestNetwork"
          >
            <div
              class="sc-bdvvtL qGXhc"
              height="20"
              width="20"
            >
              <svg>
                code.svg
              </svg>
            </div>
          </div>
        </div>
      </button>
    </div>
  </div>
  ,
</div>
`;

exports[`Wallet Card should render with wallet data and optional icon 1`] = `
<div>
  <div
    class="w-64 inline-block"
    data-testid="WalletCard__D8rr7B1d6TL6pf14LgMz4sKp1VBMs6YUYD"
  >
    <div
      class="h-48"
    >
      <button
        class="Cardstyles__CardButton-sc-ajtauu-0 dZiOWL"
        data-testid="Card"
        type="button"
      >
        <div
          class="flex relative flex-col justify-between p-5 h-full"
        >
          <div
            class="flex items-center space-x-4"
          >
            <div
              class="-space-x-1 whitespace-nowrap"
            >
              <div
                aria-label="ARK Devnet"
                class="Circle__CircleWrapper-sc-d5u41q-0 egSsGQ text-theme-secondary-700 border-theme-secondary-300 dark:border-theme-secondary-700 ring-theme-background"
                data-testid="NetworkIcon-ARK-ark.devnet"
              >
                <div
                  class="sc-bdvvtL qGXhc"
                  data-testid="NetworkIcon__icon"
                  height="20"
                  width="20"
                >
                  <svg>
                    ark.svg
                  </svg>
                </div>
              </div>
              <div
                class="Avatar__AvatarWrapper-sc-18vw8a4-0 ewMeZp flex-shrink-0"
                data-testid="Avatar"
              >
                <div
                  class="w-full h-full inline-flex items-center justify-center overflow-hidden align-middle rounded-full"
                >
                  <img
                    alt="D8rr7B1d6TL6pf14LgMz4sKp1VBMs6YUYD"
                    src="data:image/svg+xml;utf8,<svg version=\\"1.1\\" xmlns=\\"http://www.w3.org/2000/svg\\" class=\\"picasso\\" width=\\"100\\" height=\\"100\\" viewBox=\\"0 0 100 100\\"><style>.picasso circle{mix-blend-mode:soft-light;}</style><rect fill=\\"rgb(0, 150, 136)\\" width=\\"100\\" height=\\"100\\"/><circle r=\\"50\\" cx=\\"80\\" cy=\\"50\\" fill=\\"rgb(76, 175, 80)\\"/><circle r=\\"35\\" cx=\\"100\\" cy=\\"40\\" fill=\\"rgb(156, 39, 176)\\"/><circle r=\\"40\\" cx=\\"90\\" cy=\\"70\\" fill=\\"rgb(3, 169, 244)\\"/></svg>"
                    title="D8rr7B1d6TL6pf14LgMz4sKp1VBMs6YUYD"
                  />
                </div>
              </div>
            </div>
            <span
              class="font-semibold truncate text-theme-secondary-text"
            >
              My wallet
            </span>
          </div>
          <span
            class="whitespace-nowrap mt-auto text-lg font-bold text-theme-text"
            data-testid="Amount"
          >
            33.75089801 DARK
          </span>
          <span
            class="inline-flex overflow-hidden mt-1 text-xs font-semibold whitespace-nowrap text-theme-secondary-text no-ligatures"
          >
            D8rr7B1d6TL6pf14LgMz4sKp1VBMs6YUYD
          </span>
        </div>
        <div
          class="flex absolute -top-1 -right-1 items-center m-4 space-x-1"
        >
          <div
            class="inline-block p-1 text-theme-secondary-700"
            data-testid="WalletIcon__Ledger"
          >
            <div
              class="sc-bdvvtL qGXhc"
              height="20"
              width="20"
            >
              <svg>
                ledger.svg
              </svg>
            </div>
          </div>
          <div
            class="inline-block p-1 text-theme-warning-400"
            data-testid="WalletIcon__Starred"
          >
            <div
              class="sc-bdvvtL qGXhc"
              height="20"
              width="20"
            >
              <svg>
                star-filled.svg
              </svg>
            </div>
          </div>
          <div
            class="inline-block p-1 text-theme-secondary-700"
            data-testid="WalletIcon__Multisignature"
          >
            <div
              class="sc-bdvvtL qGXhc"
              height="20"
              width="20"
            >
              <svg>
                multisignature.svg
              </svg>
            </div>
          </div>
          <div
            class="inline-block p-1 text-theme-secondary-700"
            data-testid="WalletIcon__TestNetwork"
          >
            <div
              class="sc-bdvvtL qGXhc"
              height="20"
              width="20"
            >
              <svg>
                code.svg
              </svg>
            </div>
          </div>
        </div>
      </button>
    </div>
  </div>
</div>
`;
